### PR TITLE
[FEATURE] Micro run with Live-Reloading

### PR DESCRIPTION
**Pull request by h1z3y3**
We had a discussion before here: https://github.com/micro/services/discussions/345
It's an useful feature for micro service developer.

Add two args for `micro run`:
1. `--watching`, integer, specific the detect delay for file changes
2. `--force`, boolean, if it need rebuild and restart the running service

### CODE REVIEW

File: client/cli/run/cli.go [hunk 1]
```diff
@@ -46,6 +46,15 @@ var flags = []cli.Flag{
 		Name:  "metadata",
 		Usage: "Set any metadata on the service e.g. foo=bar",
 	},
+	&cli.IntFlag{
+		Name:  "watching",
```
Comment (asim) [inline client/cli/run/cli.go:50]: Change to "watch"

File: client/cli/run/cli.go [hunk 2]
```diff
@@ -46,6 +46,15 @@ var flags = []cli.Flag{
 		Name:  "metadata",
 		Usage: "Set any metadata on the service e.g. foo=bar",
 	},
+	&cli.IntFlag{
```
Comment (asim) [inline client/cli/run/cli.go:49]: Use Bool so it's --watch=true

File: client/cli/run/cli.go [hunk 3]
```diff
@@ -46,6 +46,15 @@ var flags = []cli.Flag{
...
+		Name:  "watching",
+		Usage: `Watch the file changes of source directories, then rebuild and restart the service. 
+		e.g. watching=500 means watching delay time is 500ms, if watching=0 feature will be disabled.`,
```
Comment (asim) [inline client/cli/run/cli.go:52]: Maybe call the flag --watch_delay to separately define the delay

Comment (h1z3y3) [reply client/cli/run/cli.go]: Okay, use `--watch_delay` already

Comment (h1z3y3) [reply client/cli/run/cli.go]: Now we only watch the `*.go` and `*.proto` file and only support `delay` feature

We can add some other configurations for live-reloading in the future,
like `--watch_include` or `--watch_exclude`, etc.

But for now, I think  it's enough already.

File: client/cli/run/service.go [hunk 4]
```diff
@@ -376,48 +423,9 @@ func runService(ctx *cli.Context) error {
 	// run the service
 	err = runtime.Create(srv, opts...)
 
-	watchDelay := ctx.Int("watching")
+	watchDelay := ctx.Int("watch_delay")
 	if source.Local && watchDelay > 0 {
```
Comment (asim) [inline client/cli/run/service.go:427]: I would imagine the watcher runs if I specify --watch=true. Then if I need a watch delay I add the flag --watch_delay afterwards. So you're missing the watch flag.

**Comment by h1z3y3** [comment]
@asim  `io/fs` was introduced since Go1.16, so the checks failed.

I replace it with package `os`, and using `filepath.Walk()` to replace `filepath.WalkDir()`, but it may by little ineffcient.

https://github.com/golang/go/blob/91e782106ea465acc6a4c719081cefb690b28533/src/path/filepath/path.go#L498-L511

image

**Comment by asim** [comment]
Bump the go versions in pr.yaml and tests.yaml 

https://github.com/micro/micro/tree/master/.github/workflows

**Comment by asim** [comment]
Thanks for the contribution. Looks like a great feature.

**Pull request merged**